clippy_lints/src/static_items_large_align.rs

```diff
@@ -0,0 +1,186 @@
+                            diag.span_note(span, s.as_str());
+                        }
+                    }
+                );
+            }
+        }
+    }
+}
+
+impl StaticItemsLargeAlign {
+    /// It checks a type with the following steps:
+    /// 1. Check if the type is already visited (for a static item),
+    ///     if not, continue;
+    ///     otherwise, return `None` early.
+    /// 2. Push the type in the checked types.
+    /// 3. Pick out this type if its kind is among adt, tuple, array, ref or raw ptr to them;
+    ///     otherwise return `None`.
+    /// 4. Check if its (or its inner fields') alignment are larger than page size.
+    /// 5. Return one of them;
+    ///     otherwise pop the current checked type and return `None`.
+    fn check_ty_alignment<'tcx>(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        visited_tys: &mut FxHashSet<Ty<'tcx>>,
+        intermediate_tys: &mut Vec<IntermediateTy<'tcx>>,
+    ) -> Option<LargeAlignmentTy<'tcx>> {
+        if visited_tys.contains(&ty) {
+            return None;
+        }
+        visited_tys.insert(ty);
+        intermediate_tys.push(IntermediateTy { ty });
+
+        let ret = match ty.kind() {
+            ty::Adt(adt_def, subst_ref) => {
+                if let Some(align) = adt_def.repr().align &&
+                    align.bytes() > self.page_size
+                {
+                    intermediate_tys.pop();  // the last element is already in the return value
+                    Some(LargeAlignmentTy {
+                        adt: *adt_def,
+                        name: ty.sort_string(tcx).into_owned(),
+                        align: align.bytes(),
+                    })
+                } else {
+                    adt_def.all_fields()
+                        .map(|field_ref| field_ref.ty(tcx, subst_ref))
+                        .find_map(|ty| self.check_ty_alignment(tcx, ty, visited_tys, intermediate_tys))
+                }
+            },
+            ty::Tuple(ty_list) => ty_list
+                .iter()
+                .find_map(|ty| self.check_ty_alignment(tcx, ty, visited_tys, intermediate_tys)),
+            ty::Array(ty, _) | ty::Ref(_, ty, _) | ty::RawPtr(TypeAndMut { ty, .. }) => {
+                self.check_ty_alignment(tcx, *ty, visited_tys, intermediate_tys)
+            },
+            _ => None,
+        };
+
+        if ret.is_none() {
+            intermediate_tys.pop();
+        }
+        ret
+    }
+
+    fn report_im_ty(cx: &LateContext<'_>, im_ty: &IntermediateTy<'_>) -> Option<(Span, String)> {
+        let ty = im_ty.ty;
+        if let ty::Adt(adt_def, substs_ref) = ty.kind() {
+            Some((
+                cx.tcx.def_span(adt_def.did()),
+                if substs_ref.is_empty() {
+                    format!("{} contains an inner type with large alignment", ty.sort_string(cx.tcx))
+                } else {
+                    // TODO - can we use :#?
```

Comment:
Can we? Do you have a test for this? This TODO comment should be resolved before merging.